pymeasure/instruments/attocube/adapters.py

```diff
@@ -0,0 +1,104 @@
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import time
+
+from pymeasure.adapters import TelnetAdapter
+
+
+class AttocubeConsoleAdapter(TelnetAdapter):
+    """ Adapter class for connecting to the Attocube Standard Console. This
+    console is a Telnet prompt with password authentication.
+
+    :param host: host address of the instrument
+    :param port: TCPIP port
+    :param passwd: password required to open the connection
+    :param kwargs: Any valid key-word argument for TelnetAdapter
+    """
+    def __init__(self, host, port, passwd, **kwargs):
+        self.terminator = '\r\n'
+        super(AttocubeConsoleAdapter, self).__init__(host, port, **kwargs)
+        # clear the initial messages from the controller
+        time.sleep(self.query_delay)
+        super(AttocubeConsoleAdapter, self).read()
+        # send password and check authorization
+        self.write(passwd, checkAck=False)
+        time.sleep(self.query_delay)
+        ret = super(AttocubeConsoleAdapter, self).read()
+        authmsg = ret.split(self.terminator)[1]
+        if authmsg != 'Authorization success':
+            raise Exception("Attocube authorization failed ('%s')" % authmsg)
+        # switch console echo off
+        self.ask('echo off')
+
+    def check_acknowledgement(self, reply, msg=""):
+        """ Reads the reply of the instrument until the termination sequence.
+        The returned string is check to be 'OK' and otherwise a ValueError is
+        raised.
+
+        :param msg: optional message for the eventual error
+        :returns: String ASCII response of the instrument.
+        """
+        if reply.strip() != 'OK':
+            if msg == "":  # clear buffer
+                msg = reply.strip()
+                super(AttocubeConsoleAdapter, self).read()
+            raise ValueError("AttocubeConsoleAdapter: Error after command '%s'"
+                             " with message '%s'" % (self.lastcommand, msg))
+
+    def read(self):
+        """ Reads a reply of the instrument which consists of two or more
```

Comment:
not sure if this would not be better handled by using `check_errors=True` in the property definitions?

Reply:
Ok, here I see a limit of check_errors: It is not clear to the read method how many lines the reply is. And its the last line which contains the 'OK' or 'ERROR' which would need to be checked by the check_errors function. So my approach was to read everything from the interface, check the last line for the error and declare the rest as the reply. Can this be mimicked with check_error?

Also every set command returns 'OK' or 'ERROR' which in this case could of course be easily checked by check_errors. But when the read method can not I guess it would be better to keep this in the adapter. Please advise!

Reply:
Maybe you could mimic this kind of error checking by (temporarily) storing the last response in the adapter such that it can be checked by the check_errors function? (And it would be good to remove the temporarily stored response is cleared after the error checking).

If this is, however, an adapter-wide behavior for all commands, i'm inclined to say that it would be best to keep it in the adapter.